### PR TITLE
Fix CP node isolation detection via peer reachability gate

### PR DESCRIPTION
**Pull request by mpryc**
## Summary

Fixes [RHWA-1384](https://redhat.atlassian.net/browse/RHWA-1384) — SNR agent cannot remediate isolated control plane nodes.

On CP nodes, the ClusterIP request can be routed to the local kube-apiserver endpoint. The `/readyz` checks do not provide a signal that detects loss of connectivity to the rest of the cluster, so the error count never increments, peer consultation is never triggered, and the watchdog is fed indefinitely.

### Changes

- **Peer reachability gate on CP nodes**: after readyz passes on a CP node, verify that at least one peer is reachable via lightweight TCP dial against a random sample of up to 3 peers. If no peer is reachable, treat the check as failed and enter the existing `isConsideredHealthy()` error threshold / peer consultation flow. Worker nodes are unaffected.
- **`PeerAddressProvider` interface**: changed `ApiConnectivityCheckConfig.Peers` from `*peers.Peers` to an interface. `*peers.Peers` satisfies it without modification — no changes needed in `cmd/main.go` or any consumer.
- **`PeerDialer` interface**: abstracts TCP dialing for testability.

### What does NOT change

- `/readyz?exclude=shutdown` endpoint and how it is called
- Worker node behavior
- `getWorkerPeersResponse()`, `getControlPlanePeersStatus()`, `IsControlPlaneHealthy()`, `isDiagnosticsPassed()`
- Watchdog feeding/stopping mechanics
- All existing configuration (no new CRD fields)
- `SafeTimeToAssumeNodeRebootedSeconds` calculation
- The OLM bundle

### Design

See `docs/design/cp-isolation-detection.md` for the full design document including root cause analysis, considered alternatives, edge cases, and time-to-remediation analysis.

## Test plan

- [x] 30 unit tests covering: `canReachAnyPeer`, `getAllPeerAddresses`, `isControlPlane`, full flow (CP healthy, CP isolated, worker unchanged, transient recovery), 2+1 arbiter cluster scenarios, compact cluster deduplication, interface compatibility, thread safety
- [x] All tests pass with `-race` flag
- [x] E2E: deployed on a 3x3 OCP cluster, isolated a CP node via EC2 security groups, verified remediation triggered (node rebooted, reached Fencing-Completed, recovered after isolation reverted) — [test notes](https://hackmd.io/@medik8s/SkB23VbBMg) by @weshayutin

### CODE REVIEW

File: docs/design/cp-isolation-detection.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+connectivity, kube-proxy routes the ClusterIP request to the only API server
+it can still reach: the local one running on the same node. That local API
+server returns HTTP 200 because the `/readyz` sub-checks (informer sync,
+post-start hooks) do not verify etcd quorum. The result: `errorCount` never
+increments, peer consultation is never triggered, the watchdog is fed
+indefinitely, and the node never reboots.
+
+Switching to `/healthz` does not help. The etcd healthz check tests local
+etcd connectivity, not quorum — it would still pass on an isolated node.
+
```
Comment (DanielFroehlich) [inline docs/design/cp-isolation-detection.md:18]: That is not correct. See test result in this [comment](https://redhat.atlassian.net/browse/RHWA-1384?focusedCommentId=17702302), showin health endpoint failing on isolated node with etcd check.

File: docs/design/cp-isolation-detection.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+### Why This Matters
+
+In a 2+1 arbiter topology (or any cluster with 3+ CP nodes), a
+network-isolated CP node holds its etcd member hostage. The remaining CP nodes
+cannot form quorum until the isolated node's etcd member is removed or the node
+reboots. SNR exists precisely to handle this — but the bug prevents it from
```
Comment (DanielFroehlich) [inline docs/design/cp-isolation-detection.md]: That is not correct. The remaining CP node (or in 2+1 arbiter the remaining CP node plus the arbiter) do have quorum and continue to work in degraded but operative mode. There is no hostage situation. 
Which means the controlplane is active, workload is scheduled and restarted, although it is still running on the failed/isolated node.

File: docs/design/cp-isolation-detection.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+ever firing.
+
+## Root Cause
+
+The health check has a single signal: "can I reach the API server?" On CP
+nodes this signal is tautologically true because the API server is local.
+There is no second signal to distinguish "healthy and connected" from
+"healthy but isolated."
+
```
Comment (DanielFroehlich) [inline docs/design/cp-isolation-detection.md:36]: That statement is not correct. the health check does a lot more checks then just "can I reach the API server" - see all the checks done with "?verbose" output. The root cause is that the ready endpoint does not include the ectd check.

File: docs/design/cp-isolation-detection.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+## The Fix
+
+Add a second signal: **peer reachability**. On a CP node, after readyz
+returns 200, verify that at least one peer node is reachable via TCP. If no
+peer is reachable, treat the check as failed and enter the existing error
+threshold / peer consultation flow.
+
+This is the minimal change that breaks the tautology. It requires no new CRD
+fields, no configuration changes, and no modifications to worker node behavior.
+
```
Comment (DanielFroehlich) [inline docs/design/cp-isolation-detection.md:47]: The bug has a bunch of other options on how to fix those. I would love to see them discussed with pros and cons. At least somerwhere below in "rejected alterantives" that those options have been considered, and why they have been rejected.

File: docs/design/cp-isolation-detection.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+3 × (15s + 5s + ~2s) + 30s + watchdog ≈ 96s + watchdog timeout
+```
+
+This is comparable to the existing worker-node remediation timeline.
+
+## Edge Cases
+
+| Scenario | Outcome |
+|----------|---------|
+| Single CP node, no peers | `canReachAnyPeer()` → true (no peers = cannot determine isolation), error count resets |
```
Comment (DanielFroehlich) [inline docs/design/cp-isolation-detection.md]: this includes single CP but multiple worker node topologies i guess? Would love to see that stated explicitly, to make it clear that case has been considered.

File: docs/design/cp-isolation-detection.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+when the error threshold is reached.
+
+**2. CP isolation escalation** (`isConsideredHealthy()`)
+
+A secondary issue: when readyz *does* fail on a CP node (e.g., iptables DROP
+rather than NIC down), the existing code path enters `isConsideredHealthy()`
+but the CP peer isolation signal can be discarded if the worker error threshold
+has not yet been reached. This delays remediation unnecessarily.
+
+A new counter (`cpUnreachableCount`) tracks consecutive CP peer unreachability
```
Comment (slintes) [inline docs/design/cp-isolation-detection.md]: I don't understand how this change improves anything. You replace one error treshold with another one, not? IMHO this adds complexity with low value if any.

Comment (mpryc) [reply docs/design/cp-isolation-detection.md]: You are right. I'll drop second counter. However @DanielFroehlich  shared test data showing /healthz already catches isolation via the etcd sub-check. If /livez does too, the fix might be as simple as switching the endpoint. In that case it would be almost a one-liner. wdyt ?

Comment (slintes) [reply docs/design/cp-isolation-detection.md]: see https://redhat.atlassian.net/browse/RHWA-1384?focusedCommentId=17702903

Comment (slintes) [reply docs/design/cp-isolation-detection.md]: My initial fav fix would be to not use the cluster ip for the API check, but the node IP of another CP node. 
But even that needs further considerations... what if the other node is down, how does peer check behave with that change, .... it's too easy to break things.

Always keep in mind why we do what we do:

- we want to reboot when a SNR CR is created
- that only works when we have a working API server connection
- that's why we introduced the API check: ensure we can watch for SNR CRs
- what if that connection is broken:
  - is it a problem with the API server? If so, nothing we can do
  - is it a problem with our own node? Reboot, the manager pod cares for everything else
  - in order to find out which of these 2 apply, we ask our peers

I think your proposed fix has the least chance for unwanted side effects...

File: internal/apicheck/check.go [hunk 7]
```diff
@@ -110,14 +139,98 @@ func (c *ApiConnectivityCheck) Start(ctx context.Context) error {
...
+			combined = append(combined, p)
+		}
+	}
+	for _, p := range cpPeers {
+		if _, exists := seen[p.IP]; !exists && p.IP != "" {
+			seen[p.IP] = struct{}{}
+			combined = append(combined, p)
+		}
+	}
+	return combined
```
Comment (slintes) [inline internal/apicheck/check.go:234]: you might want to shuffle peers... a deterministic first N selection could false-positive on larger clusters if first 3 peers are down

File: docs/design/cp-isolation-detection.md (hunk 6 above)
Comment (DanielFroehlich) [reply docs/design/cp-isolation-detection.md]: ```we want to reboot when a SNR CR is created``
That is a design flaw: On a network isolated node, there is no way to detect the SNR CR being created.
SNR can not depend on a failed node to detect that CR to be created.

There is also no way to contact peers. That is if API is detected faulty (e.g. via health endpoint), a peer check should also fail on the isolated node (connection timeout). If the peer check does not fail if it cant connect to any peer, that would also be a bug / design flaw.

**Comment by weshayutin** [comment]
OK..  I have successfully recreated the issue AFAIK and to the best of my current ability.   Using ec2 security groups I completely isolated the networking of one the master nodes in a typical 3x3 cluster.   As documented in the bug that @DanielFroehlich opened, SelfNodeRemediation instance was created and no action is taken.. that's a bug imho.  Notes are here: https://hackmd.io/@medik8s/HyQdnN-Bzg    

Now today I pulled in @mpryc's fix and deployed it to the same cluster.  and ran the same scenario after ensuring all nodes were running etc..  This time a selfnoderemediation instance was created, the node rebooted and ended up in "Fencing-Complete"  .. I am not an expert but this seems right.. 

Then I reverted the change to the ec2 security groups and the node went back to "Ready" and the selfnoderemediation instance was deleted.   These notes are here: https://hackmd.io/@medik8s/SkB23VbBMg

The cr transitions can be seen here: 

```
oc get selfnoderemediations 
NAME                                              AGE
ip-10-0-76-226.us-west-2.compute.internal-6z94q   2m
whayutin@fedora:~/OPENSHIFT/git/MEDIK8S/node-healthcheck-operator$ oc get selfnoderemediations ip-10-0-76-226.us-west-2.compute.internal-6z94q -o yaml
apiVersion: self-node-remediation.medik8s.io/v1alpha1
kind: SelfNodeRemediation
metadata:
  annotations:
    remediation.medik8s.io/node-name: ip-10-0-76-226.us-west-2.compute.internal
    remediation.medik8s.io/template-name: self-node-remediation-automatic-strategy-template
  creationTimestamp: "2026-07-24T19:14:05Z"
  finalizers:
  - self-node-remediation.medik8s.io/snr-finalizer
  generateName: ip-10-0-76-226.us-west-2.compute.internal-
  generation: 1
  labels:
    app.kubernetes.io/part-of: node-healthcheck-controller
    remediation.medik8s.io/isControlPlaneNode: ""
  name: ip-10-0-76-226.us-west-2.compute.internal-6z94q
  namespace: openshift-workload-availability
  ownerReferences:
  - apiVersion: remediation.medik8s.io/v1alpha1
    controller: false
    kind: NodeHealthCheck
    name: nhc-control-plane
    uid: ba63397c-fb4b-4c1c-90a5-84234275180e
  resourceVersion: "96809"
  uid: 78cb0ce3-65b5-4c3d-8465-58338084796a
spec:
  remediationStrategy: Automatic
status:
  conditions:
  - lastTransitionTime: "2026-07-24T19:14:05Z"
    message: ""
    reason: RemediationStarted
    status: "True"
    type: Processing
  - lastTransitionTime: "2026-07-24T19:14:05Z"
    message: ""
    reason: RemediationStarted
    status: Unknown
    type: Succeeded
  phase: Reboot-Completed
  timeAssumedRebooted: "2026-07-24T19:16:05Z"
whayutin@fedora:~/OPENSHIFT/git/MEDIK8S/node-healthcheck-operator$ oc get selfnoderemediations ip-10-0-76-226.us-west-2.compute.internal-6z94q -o yaml
apiVersion: self-node-remediation.medik8s.io/v1alpha1
kind: SelfNodeRemediation
metadata:
  annotations:
    remediation.medik8s.io/node-name: ip-10-0-76-226.us-west-2.compute.internal
    remediation.medik8s.io/template-name: self-node-remediation-automatic-strategy-template
  creationTimestamp: "2026-07-24T19:14:05Z"
  finalizers:
  - self-node-remediation.medik8s.io/snr-finalizer
  generateName: ip-10-0-76-226.us-west-2.compute.internal-
  generation: 1
  labels:
    app.kubernetes.io/part-of: node-healthcheck-controller
    remediation.medik8s.io/isControlPlaneNode: ""
  name: ip-10-0-76-226.us-west-2.compute.internal-6z94q
  namespace: openshift-workload-availability
  ownerReferences:
  - apiVersion: remediation.medik8s.io/v1alpha1
    controller: false
    kind: NodeHealthCheck
    name: nhc-control-plane
    uid: ba63397c-fb4b-4c1c-90a5-84234275180e
  resourceVersion: "97147"
  uid: 78cb0ce3-65b5-4c3d-8465-58338084796a
spec:
  remediationStrategy: Automatic
status:
  conditions:
  - lastTransitionTime: "2026-07-24T19:16:26Z"
    message: ""
    reason: RemediationFinishedSuccessfully
    status: "False"
    type: Processing
  - lastTransitionTime: "2026-07-24T19:16:26Z"
    message: ""
    reason: RemediationFinishedSuccessfully
    status: "True"
    type: Succeeded
  phase: Fencing-Completed
  timeAssumedRebooted: "2026-07-24T19:16:05Z"
whayutin@fedora:~/OPENSHIFT/git/MEDIK8S/node-healthcheck-operator$ oc get selfnoderemediations ip-10-0-76-226.us-west-2.compute.internal-6z94q -o yaml
apiVersion: self-node-remediation.medik8s.io/v1alpha1
kind: SelfNodeRemediation
metadata:
  annotations:
    remediation.medik8s.io/node-name: ip-10-0-76-226.us-west-2.compute.internal
    remediation.medik8s.io/template-name: self-node-remediation-automatic-strategy-template
  creationTimestamp: "2026-07-24T19:14:05Z"
  finalizers:
  - self-node-remediation.medik8s.io/snr-finalizer
  generateName: ip-10-0-76-226.us-west-2.compute.internal-
  generation: 1
  labels:
    app.kubernetes.io/part-of: node-healthcheck-controller
    remediation.medik8s.io/isControlPlaneNode: ""
  name: ip-10-0-76-226.us-west-2.compute.internal-6z94q
  namespace: openshift-workload-availability
  ownerReferences:
  - apiVersion: remediation.medik8s.io/v1alpha1
    controller: false
    kind: NodeHealthCheck
    name: nhc-control-plane
    uid: ba63397c-fb4b-4c1c-90a5-84234275180e
  resourceVersion: "97147"
  uid: 78cb0ce3-65b5-4c3d-8465-58338084796a
spec:
  remediationStrategy: Automatic
status:
  conditions:
  - lastTransitionTime: "2026-07-24T19:16:26Z"
    message: ""
    reason: RemediationFinishedSuccessfully
    status: "False"
    type: Processing
  - lastTransitionTime: "2026-07-24T19:16:26Z"
    message: ""
    reason: RemediationFinishedSuccessfully
    status: "True"
    type: Succeeded
  phase: Fencing-Completed
  timeAssumedRebooted: "2026-07-24T19:16:05Z"

```

@slintes am I correct that your opinion is that the node should NOT reboot. which I think is reasonable as it didn't fix anything in this case ( networking issues ).  What is your suggested remediation, as I don't really know :) .. Thanks  in advance.

File: docs/design/cp-isolation-detection.md (hunk 6 above)
Comment (slintes) [reply docs/design/cp-isolation-detection.md]: What I wrote is a a very basic flow without going into technical details. I don't know if I explained it so badly, but I'm pretty sure there is no design flaw...

>  On a network isolated node, there is no way to detect the SNR CR being created.

that´s why we have all the other steps

> If the peer check does not fail if it can't connect to any peer

of course it does

**Comment by slintes** [comment]
> @slintes am I correct that your opinion is that the node should NOT reboot

No, where do you see that? An isolated node needs to be fenced so that workloads can be rescheduled safely. Fencing happens by reboot. If the reboot fixes the network issue or not is not important. A reboot fixing any issues is a very welcome side effect but not the primary goal of what we do.

File: docs/design/cp-isolation-detection.md (hunk 1 above)
Comment (mpryc) [reply docs/design/cp-isolation-detection.md:18]: Fixed. The problem statement now says the /readyz checks "do not provide a signal that detects loss of connectivity to the rest of the cluster" - no claim about whether /readyz includes or excludes specific etcd sub-checks. Thanks for the test data showing /healthz catches isolation via the etcd sub-check.

it's  also documented in the "Considered Alternatives" section. See also the API health endpoints docs for reference on the available sub-checks.

File: docs/design/cp-isolation-detection.md (hunk 2 above)
Comment (mpryc) [reply docs/design/cp-isolation-detection.md]: Fixed. The section now correctly states that remaining nodes maintain etcd quorum and the cluster continues operating in degraded mode.

File: docs/design/cp-isolation-detection.md (hunk 3 above)
Comment (mpryc) [reply docs/design/cp-isolation-detection.md:36]: Fixed. The root cause now describes that /readyz runs multiple sub-checks but does not provide a signal that detects loss of connectivity to the rest of the cluster. Link to k8s sub-checks https://kubernetes.io/docs/reference/using-api/health-checks/

File: docs/design/cp-isolation-detection.md (hunk 4 above)
Comment (mpryc) [reply docs/design/cp-isolation-detection.md:47]: Added three "Considered Alternatives", please review if I am covering them correctly and if we want to consider more options. I think it's always good to have all of freshly brainstormed in case in the future one will want to modify behavior.

File: docs/design/cp-isolation-detection.md (hunk 6 above)
Comment (mpryc) [reply docs/design/cp-isolation-detection.md]: Dropped. The CP gate now enters the existing isConsideredHealthy() flow directly.

File: internal/apicheck/check.go (hunk 7 above)
Comment (mpryc) [reply internal/apicheck/check.go:234]: Done. [Added rand.Shuffle.](https://github.com/medik8s/self-node-remediation/pull/325/changes#diff-bb88ec6331507bace73a5ee08b0c826b96195b08898305399806110bb80f563aR186)

File: internal/apicheck/check.go [hunk 8]
```diff
@@ -32,16 +33,43 @@ import (
 
 const (
 	eventReasonPeerTimeoutAdjusted = "PeerTimeoutAdjusted"
+
+	peerReachabilityTimeout = 2 * time.Second
+	maxPeersToSample        = 3
 )
```
[the remark this answers is not in the document]
Comment (mpryc) [reply internal/apicheck/check.go:40]: intentional design decision. peerReachabilityTimeout (2s, TCP SYN check) is deliberately separate from config.PeerDialTimeout (gRPC health dial). They serve different purposes:
- peerReachabilityTimeout: lightweight "is the network up?" check - should always be fast
- PeerDialTimeout: full gRPC peer health connection - may be tuned higher for slow networks

Merging them would mean operators tuning gRPC timeouts unintentionally slow down isolation detection. The design doc uses the matching name peerReachabilityTimeout=2s to avoid confusion. No change needed imo. Please let me know if you agree ?

File: docs/design/cp-isolation-detection.md (hunk 5 above)
Comment (mpryc) [reply docs/design/cp-isolation-detection.md]: yes, single CP + multiple workers is now explicitly called out in the Edge Cases table. Two separate rows: "Single CP node, no peers (e.g. SNO)" and "Single CP node + multiple workers" - both state no false-positive self-reboot risk, with the mechanism explained (getAllPeerAddresses() includes worker peers, MinPeersForRemediation safety threshold).

**Comment by mpryc** [comment]
Please review this PR and let me know if we all agree on the current design doc and implementation changes or this require further updates/changes.

**Comment by mpryc** [comment]
Closing, because we have now: https://github.com/medik8s/self-node-remediation/pull/339/
